### PR TITLE
Undo and redo functionality 📜

### PR DESCRIPTION
**Pull request by MusicFreak456**
Implement undo and redo functionality for following events:
- [x] Shape drawing / deletion
- [x] Shape update (eg. color change)
- [x] Shape correction / handwriting recognition

### CODE REVIEW

File: src/store/CanvasStore.ts [hunk 1]
```diff
@@ -4,49 +4,110 @@ import { ref, type Ref } from 'vue'
...
-  function removeDrawnShapeById(id: string) {
+  function removeDrawnShapeById(id: string, record = true) {
     const shapeIndex = drawnShapes.value.findIndex((shape) => shape.id == id)
     if (shapeIndex < 0) return
+    if (record) recordShapeDeletion(drawnShapes.value[shapeIndex])
     drawnShapes.value.splice(shapeIndex, 1)
-    IoConnection.emit('shape_delete', {
-      board_id: boardStore.boardId,
-      shape_id: id
-    })
```
Comment (ventus550) [inline src/store/CanvasStore.ts:31]: What happens to this?

File: src/services/history/HistoryService.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        this.undoShapesReplacementEvent(event)
+        break
+      case HistoryEventType.SHAPE_UPDATED:
+        this.undoShapeUpdatedEvent(event)
+        break
+      case HistoryEventType.AGGREGATE:
+        this.undoAggregateEvent(event)
+        break
+    }
+  }
```
Comment (ventus550) [inline src/services/history/HistoryService.ts:46]: This whole event bussiness thing feels like it deserves its own class hierachy. Not sure how that would work out in practice though...

File: src/store/CanvasStore.ts (hunk 1 above)
Comment (MusicFreak456) [reply src/store/CanvasStore.ts:31]: Moved to separate function: commitShapeDeletion

File: src/services/history/HistoryService.ts (hunk 2 above)
Comment (MusicFreak456) [reply src/services/history/HistoryService.ts:46]: Not sure what you mean by that. Each event has more or less unique set of information that it needs to save to be reversable and I think it's rather convenient that TypeScript even allows to easily express such polymorphism in the first place

Comment (ventus550) [reply src/services/history/HistoryService.ts:46]: Think on that!

Comment (MusicFreak456) [reply src/services/history/HistoryService.ts:46]: I thought about about that, and you are right! I will rewrite that so events will be objects that will know how to do and redo them

**Pull request merged**